### PR TITLE
[CLIPPER-113] Implement model active version tracking and updates

### PR DESCRIPTION
**Pull request by dcrankshaw**
This PR associated a `current_model_version` with each model and makes applications specify candidate model names but not versions. Instead, the current version of a model is deployed. Model versions are automatically updated when a newer model version is deployed. I also implemented a `set_model_version` REST call in the management frontend that allows users to manually set the current model version to any version that has previously been deployed. This allows for model rollback and roll-forward.

I removed the feedback section of the selection policy and the tutorial wording still needs a little bit of work.

### CODE REVIEW

**Comment by dcrankshaw** [comment]
Note that in order to test this with the revised tutorial, you will need to build the `query_frontend` and `management_frontend` containers locally. The ones deployed to DockerHub are from the develop branch.

**Comment by dcrankshaw** [comment]
@nishadsingh1 Can you make a version of the [deploy sklearn model image](https://github.com/ucbrise/clipper/blob/develop/examples/tutorial/img/deploy_sklearn_model.png) from the tutorial for the TensorFlow model? In other words, delete the sklearn box from [this image](https://github.com/ucbrise/clipper/blob/develop/examples/tutorial/img/deploy_tf_model.png). With this new model versioning PR, the TF model _replaces_ the sklearn model and I would like to illustrate that.

Also, what did you use to make those images? Can you add whatever powerpoint/google slides file you used to the Clipper Google Drive?

**Comment by dcrankshaw** [comment]
jenkins test this please

File: src/frontends/src/query_frontend.hpp
Comment (Corey-Zumar) [inline src/frontends/src/query_frontend.hpp:164]: Do we need to hold this lock for the entire query processing and response creation procedure?

File: src/frontends/src/query_frontend.hpp
Comment (Corey-Zumar) [inline src/frontends/src/query_frontend.hpp:203]: As above, do we need to hold this lock for the entire update procedure or just until the `versioned_models` map is populated? If the answer is the latter, releasing the lock earlier may avoid resource contention between the `update` and `predict` functions.

File: src/libclipper/include/clipper/redis.hpp
Comment (Corey-Zumar) [inline src/libclipper/include/clipper/redis.hpp]: Nit: check documentation formatting

File: src/libclipper/src/metrics.cpp
Comment (Corey-Zumar) [inline src/libclipper/src/metrics.cpp:1]: Is this include necessary? (Maybe gcc is throwing an error where clang isn't?, etc)

File: src/libclipper/src/redis.cpp
Comment (Corey-Zumar) [inline src/libclipper/src/redis.cpp]: Should we verify that `result` is a nonnegative integer?

File: src/libclipper/src/redis.cpp
Comment (Corey-Zumar) [inline src/libclipper/src/redis.cpp]: Nit: For clarity, defining `const std::string NO_PREFIX = "";` and using it in place of `""` is clearer.

File: management/clipper_manager.py
Comment (Corey-Zumar) [inline management/clipper_manager.py]: We should make this the only point through which model rollback / rollforward can be achieved. Currently, we can also call `deploy_model()` while specifying a previously-added version of a model. However, `deploy_model()` has no checks in place to ensure that the model being added is actually equivalent to the old model corresponding to the given version. Hypothetically, version x of a model can therefore change throughout time, and this seems to defeat the purpose of versioning. 

We can fix this by adding checks to `deploy_model()` that fail if a model is deployed with a previously-specified version.

File: src/management/src/management_frontend_tests.cpp
Comment (Corey-Zumar) [inline src/management/src/management_frontend_tests.cpp]: Wouldn't we expect `get_current_model_version()` to return `4`, corresponding to the last successful call to `add_model()` with `model4`? Is attempting to set an invalid model version supposed to change the model version to `-1`? If so, we should consistently maintain the previous version in the face of an update failure, if possible.

File: management/clipper_manager.py
Comment (Corey-Zumar) [inline management/clipper_manager.py]: Missing `self` in parameter list

**Review by Corey-Zumar — CHANGES_REQUESTED** [review]
Tests pass and functionality seems to be mostly complete. Comments address a couple of concerns involving behavior with `deploy_model()` in conjunction with `set_model_version`. Additionally:

* Line 491 of clipper_manager.py should refer to `Clipper.deploy_model()` instead of `Clipper.add_model()`.

* What’s the intent of the “Warning” image in the tutorial_part_2 notebook? Seems unclear to me.

File: src/frontends/src/query_frontend.hpp
Comment (dcrankshaw) [reply src/frontends/src/query_frontend.hpp:164]: Nope. The lock is in a nested block so it will go out of scope and be released on line 171 before the prediction is made.

File: src/frontends/src/query_frontend.hpp
Comment (dcrankshaw) [reply src/frontends/src/query_frontend.hpp:203]: The lock goes out of scope on line 210.

File: src/libclipper/include/clipper/redis.hpp
Comment (dcrankshaw) [reply src/libclipper/include/clipper/redis.hpp]: fixed

File: src/libclipper/src/metrics.cpp
Comment (dcrankshaw) [reply src/libclipper/src/metrics.cpp:1]: `std::sort` which we use in this file is defined in `<algorithm>`. I prefer to explicitly include the header dependencies for a file as much as possible and I noticed this one was missing.

File: src/libclipper/src/redis.cpp
Comment (dcrankshaw) [reply src/libclipper/src/redis.cpp]: Yeah good idea.

File: src/libclipper/src/redis.cpp
Comment (dcrankshaw) [reply src/libclipper/src/redis.cpp]: Hmm I actually disagree here. Defining NO_PREFIX as a constant seems like treating this as a special case. I'll define a local variable `std::string prefix = ""` in each method instead.

File: management/clipper_manager.py
Comment (dcrankshaw) [reply management/clipper_manager.py]: Yeah this is a really good point.

Comment (dcrankshaw) [reply management/clipper_manager.py]: Fixed

File: src/management/src/management_frontend_tests.cpp
Comment (dcrankshaw) [reply src/management/src/management_frontend_tests.cpp]: I was adding the model directly to Redis so `set_model_version` wasn't being called. I fixed this by calling the managament frontend's `RequestHandler::add_model` with JSON and it now should reflect the expected behavior.

File: management/clipper_manager.py
Comment (dcrankshaw) [reply management/clipper_manager.py]: I added checks to the management frontend's `add_application` and `add_model` methods to reject duplicate adds.

**Comment by dcrankshaw** [comment]
@Corey-Zumar I addressed your comments. I also added a check to prevent duplicate adding of applications or model versions. This fixes #112.

**Review by Corey-Zumar — APPROVED** [review]
Awesome, LGTM. Will merge after jenkins is able to build successfully.

**Comment by dcrankshaw** [comment]
@Corey-Zumar BTW the warning in the tutorial is because @nishadsingh1 said he accidentally ran cleanup too early the first time he ran the tutorial. The idea was to have a visual separation from the tutorial code and the cleanup. It's been there for awhile, I just moved the location around slightly.

**Comment by dcrankshaw** [comment]
@Corey-Zumar Hold off on merging this until the tutorial is updated. Should be sometime today.

**Comment by dcrankshaw** [comment]
@Corey-Zumar Tutorial is updated. This PR is good to go once Jenkins passes it.

**Pull request merged**
